### ISSUE TITLE
Differences between plyr functions and functions in compat-plyr.R are causing failures

### ISSUE THREAD

**Issue by paleolimbot** [tidyverse/ggplot2#3315]
This is a revdep issue (#3303) in which the functions in `compat-plyr.R` are returning slightly different output than their former **plyr** counterparts, which in turn cause package failures.

- Different output column ordering between `plyr::ddply` and `ggplot2:::dapply` results in a different order of evaluation, which causes random jitters to be slightly different between **ggplot2** and **ggstance** (which is still using plyr). **gstance** tests for consistency between `position_jitterdodge()` and `position_jitterdodgev()`, which is why this results in a revdep failure.

``` r
df <- data.frame(y = 1:4, by_col = c("a", "b"))
plyr::ddply(df, "by_col", identity)
#>   y by_col
#> 1 1      a
#> 2 3      a
#> 3 2      b
#> 4 4      b
ggplot2:::dapply(df, "by_col", identity)
#>   by_col y
#> 1      a 1
#> 2      a 3
#> 3      b 2
#> 4      b 4
```

- Different (reversed order of) type coercion  in `rbind_dfs()` is causing the behaviour of `join_keys()` to differ from `plyr::join.keys()`. This results in the failure of faceting by some S3 vectors (#3313).

``` r
df <- data.frame(date = as.Date("2019-01-01"))
df_fct <- data.frame(date = as.factor(df$date))
unlist(ggplot2:::join_keys(df, df_fct, "date"))
#> x y n 
#> 1 1 1
unlist(ggplot2:::join_keys(df_fct, df, "date"))
#> x y n 
#> 2 1 2

unlist(plyr::join.keys(df, df_fct, "date"))
#> x y n 
#> 2 1 2
unlist(plyr::join.keys(df_fct, df, "date"))
#> x y n 
#> 1 1 1
```

**Comment by clauswilke** [comment]
@thomasp85 Is it possible to rewrite the internal ggplot2 functions so they mirror the output from the plyr functions? If it's possible without too much effort then I think we should do it.

**Comment by thomasp85** [comment]
It should be (and I thought it was before the revdepchecks)

**Comment by thomasp85** [comment]
closed by #3327
